include/fiction/algorithms/simulation/sidb/displacement_robustness_domain.hpp

```diff
@@ -0,0 +1,670 @@
+            std::max(uint64_t{1},
+                     static_cast<uint64_t>(static_cast<double>(all_combinations_of_fabricating_misplaced_sidbs.size()) *
+                                           std::min(params.percentage_of_analyzed_displaced_layouts, 1.0)));
+
+        uint64_t number_of_tested_misplaced_cell_combinations = 0;
+
+        for (const auto& fixed_c_indices : all_combinations_of_fabricating_misplaced_sidbs)
+        {
+            if (number_of_tested_misplaced_cell_combinations >= number_of_maximal_tested_misplaced_cell_combinations)
+            {
+                break;
+            }
+            // all cells are fixed initially.
+            for (const auto& c : sidbs_of_the_original_layout)
+            {
+                params.fixed_sidbs.insert(c);
+            }
+            for (const auto& c : fixed_c_indices)
+            {
+                const auto cells = sidbs_of_the_original_layout[c];
+                params.fixed_sidbs.erase(cells);
+            }
+
+            determine_robustness_domain();
+
+            number_of_tested_misplaced_cell_combinations++;
+        }
+
+        return static_cast<double>(stats.num_operational_sidb_displacements) /
+               static_cast<double>(stats.num_non_operational_sidb_displacements +
+                                   stats.num_operational_sidb_displacements);
+    }
+
+  private:
+    /**
+     * The SiDB layout for which the displacement robustness calculation is performed.
+     */
+    const Lyt& layout;
+    /**
+     * The parameters for the displacement robustness computation.
+     */
+    displacement_robustness_domain_params<cell<Lyt>> params;
+    /**
+     * The statistics of the displacement robustness computation.
+     */
+    displacement_robustness_domain_stats& stats;
+    /**
+     * This stores all possible displacements for all SiDBs in the SiDB layout. This means e.g. the first vector
+     * describes all possible positions of the first SiDB due to the allowed/possible displacements.
+     */
+    std::vector<std::vector<cell<Lyt>>> all_possible_sidb_displacements{};
+    /**
+     * SiDB positions of the originally given SiDB layout.
+     */
+    std::vector<cell<Lyt>> sidbs_of_the_original_layout{};
+    /**
+     * The logical specification of the layout.
+     */
+    const std::vector<TT>& truth_table;
+    /**
+     * Random device for obtaining seed for the random number generator.
+     * Provides a source of quasi-non-deterministic pseudo-random numbers.
+     */
+    std::random_device rd{};
+    /**
+     * Mersenne Twister random number generator.
+     * Generates high-quality pseudo-random numbers using a random seed from 'rd'.
+     */
+    std::mt19937 generator;
+    /**
+     * This function calculates all permitted displacements for each SiDB based on the specified allowed displacements.
+     */
+    void calculate_all_possible_displacements_for_each_sidb() noexcept
```

Comment:
This function got quite large. Can we break it down?

Reply:
I know what you mean. I thought about it for a long time, and I have to say that it looks strange to have a separate function that is only called when a certain coordinate type is present. I would rather leave it as it is. Or do you have an easy way how to split it?